Countries, currencies and regions infrastructure update

JeffPyeBrook opened the pull request:
Update countries, regions and currencies to be PHP code based rather than from the database.
- Allows internationalization of country, region and currency information
- Fully backwards compatible with 
- Eliminates a significant amount of data being read from the database at startup
- Removes several hundred lines of code from what we have to maintain
- also improves speed and lowers memory footprint

This early PR is being made available so that the peer review can start now while work is being completed

JeffPyeBrook committed:
create country init routine

JeffPyeBrook committed:
Update countries, regions and currencies to be PHP code based rather than from the database.
Allows internationalization of country, region and currency information
Eliminates a significant amount of data being read from the database at startup
Removes several hundred lines of code from what we have to maintain
and also improves speed and lowers memory footprint

JeffPyeBrook commented:
Items in progress
* Integration testing
* Comparison and data management routine to pick up changes to core country data ( visibility already works )
* Additional filters for plugin integrators

JeffPyeBrook commented:
@leewillis77 @JustinSainton this is working in my test environment, i'll keep working on it, but it is reviewable in its current state

JeffPyeBrook commented on wpsc-includes/country-api.php:
```diff
@@ -0,0 +1,193 @@
+<?php
```
@JustinSainton @leewillis77 I would like this file to contain our outward facing procedural api for countries/regions/currency operations.  dev's certainly can go right at the class interface for the WPSC_Countries class, but that has a fair amount of complexisty because of some backwards compatibility requirements.

For most simple operations that a plugin or theme might want to do this file seem to be the API we shoulf promote, in combinatiuon with the WPSC_Country and WPSC_Region objects that the functions within the file return

JustinSainton commented on wp-shopping-cart.php:
```diff
@@ -123,11 +123,9 @@ public function setup_table_names() {
 		$wpdb->wpsc_coupon_codes        = WPSC_TABLE_COUPON_CODES;
 		$wpdb->wpsc_cart_contents       = WPSC_TABLE_CART_CONTENTS;
 		$wpdb->wpsc_claimed_stock       = WPSC_TABLE_CLAIMED_STOCK;
-		$wpdb->wpsc_currency_list       = WPSC_TABLE_CURRENCY_LIST;
```
Why are these tables removed?  We talked in Slack about back compat concerns here...I might be missing something though?

JeffPyeBrook replied:
As long as deprecated is true they are still defined, and even installed if it a brand new installation.  Also, ( when I finish up the last routine ) and changes made through $wpdb calls will also be automatically picked up by the code.

At some point when we want to turn off deprecated the the "legacy" database   structures and constants will quietly go away without further action

JustinSainton replied:
It looks like the constants are defined (https://github.com/wp-e-commerce/WP-e-Commerce/pull/1865/files#diff-66e9d8bfe5d295da98d32c2d9c5a8d5cR251), but the WPDB properties are not?

JeffPyeBrook committed:
Initalize countries data onm wpsc_laoded rather than init
Use documented API function for countires operation rather than static class method

JeffPyeBrook commented on wpsc-taxes/models/taxes.class.php:
```diff
@@ -237,75 +244,35 @@ function wpec_taxes_get_countries( $visibility = 'any' ) {
+	 * @return array, int, string or false
 	 * */
 	function wpec_taxes_get_country_information( $columns = false, $where = array(), $order_by = false ) {
-		//check for all-markets
```
@JustinSainton this removal needs to be reviewed

JustinSainton replied:
Seems fine to me.

JustinSainton commented:
I'd like to see this in 4.0.  Are all unit tests still passing and is this merge-ready?

JustinSainton commented on wpsc-admin/includes/display-items-functions.php:
```diff
@@ -127,12 +127,10 @@ function wpsc_price_control_forms() {
 	/* Get country name and symbol */
-	$currency_type = get_option( 'currency_type' );
-	$country       = new WPSC_Country( $currency_type );
+	$country       = wpsc_get_currency_type_country_object();
```
I'm not a massive fan of ending function names in data types, e.g. `WPSC_Countries::get_countries_array()` or this, `wpsc_get_currency_type_country_object()`. Seems to me it'd be better to just be `wpsc_get_currency_type_country`?

JustinSainton commented on wpsc-admin/includes/save-data.functions.php:
```diff
@@ -554,13 +554,14 @@ function wpsc_save_category_set( $category_id, $tt_id ) {
 		if ( ! empty( $_POST['countrylist2'] ) && ( $category_id > 0 ) ) {
 			$AllSelected = false;
-			$countryList = $wpdb->get_col( "SELECT `id` FROM  `" . WPSC_TABLE_CURRENCY_LIST . "`" );
+			$countryList = wpsc_get_country_objects();
```
LIkewise, here, seems `wpsc_get_countries()` would make more sense.

JustinSainton commented on wpsc-includes/country-api.php:
```diff
@@ -0,0 +1,193 @@
+
+
+/**
+ * Get the country id correcposnding to the currently configured store currency type, might be the default
```
Docs need a grammar/spelling pass.

Also, this function name is sort of obtuse.  I get what it's doing, but it just feels a bit much.  Not sure what I'd change just yet, but i feel like a lot of this PR suffers from [number two](http://martinfowler.com/bliki/TwoHardThings.html)

JustinSainton commented on wpsc-includes/wpsc-countries.class.php:
```diff
@@ -745,875 +768,201 @@ public static function get_country_id_by_country_code( $country_code ) {
+	 * @param WPSC_Country $wpsc_country
 	 */
-	private static $country_code_by_country_id = null;
+	function save_country_visibility_override( $wpsc_country ) {
```
Can you elaborate a bit of what's going on here, and what exactly we're storing here in the wpsc_country_visibility_overrides option?

JustinSainton commented on wpsc-includes/wpsc-countries.class.php:
```diff
@@ -745,875 +768,201 @@ public static function get_country_id_by_country_code( $country_code ) {
+ *
+ * @return void
+ */
 function _wpsc_make_countries_data_available() {
```
This seems unnecessary. Couldn't we just have `add_action( 'wpsc_loaded', 'WPSC_Countries::get_instance' );`?

JustinSainton commented:
@JeffPyeBrook What's your bandwidth like for this for the 4.1 cycle?

JeffPyeBrook commented:
i'm around

JustinSainton commented:
Yay!  What's your sense on the ETA for the work here?

JeffPyeBrook commented:
i'll have to take a look and see what is left to be done, my vague recollection was that it was substantially done. i would guess that the code is in some of my production configurations and has been for a while?

JeffPyeBrook committed:
fix widget constructors

JustinSainton commented:
@JeffPyeBrook Let me know if you have an ETA on this - if we can get all the unit tests passing, comments addressed, and PR rebased, it would be great to get this in 4.0.
